### PR TITLE
OJ-1139 - Convert authorization lambda function to TypeScript (draft)

### PR DESCRIPTION
**Pull request by tal-nagra**
### What changed
- TBC

### Why did it change
- Conversion from `java` to `TypeScript`

### Issue tracking
- [OJ-1139](https://govukverify.atlassian.net/browse/OJ-1139)

[OJ-1139]: https://govukverify.atlassian.net/browse/OJ-1139?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy.sh [hunk 1]
```diff
@@ -33,7 +33,7 @@ echo -e "\tCriIdentifier SSM key ${GREEN}$cri_identifier${NOCOLOR}"
 
 ./gradlew clean
 sam validate -t infrastructure/lambda/template.yaml
-sam build -t infrastructure/lambda/template.yaml
+sam build -t infrastructure/lambda/template.yaml --beta-features
```
Comment (MattBidewell) [inline deploy.sh:36]: Doesn't need `--beta-features` update your SAM cli instead

File: authorization2/package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "description": "Authorization Typescript Node Lambda",
+    "main": "app.js",
+    "dependencies": {
+        "@aws-lambda-powertools/logger": "^1.5.0",
+        "@aws-lambda-powertools/metrics": "^1.5.0",
+        "@aws-sdk/client-dynamodb": "3.150.0",
+        "@aws-sdk/client-ssm": "^3.186.0",
+        "@aws-sdk/lib-dynamodb": "3.150.0",
+        "esbuild": "0.14.14",
+        "uuid": "^9.0.0"
```
Comment (MattBidewell) [inline authorization2/package.json:13]: Is there a reason we're using uuid over native crypto/uuid 

- https://nodejs.org/api/crypto.html#:~:text=n%7D%60)%3B-,crypto.randomUUID(%5Boptions%5D),-%23

Comment (merlinc) [reply authorization2/package.json:13]: Further reading: https://www.nearform.com/blog/new-crypto-capabilities-in-node-js/

File: authorization2/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
+    "compilerOptions": {
+        "experimentalDecorators": true,
+        "target": "es2020",
+        "strict": true,
+        "preserveConstEnums": true,
+        // "noEmit": true, // emit for debugger testing purposes.
+        "sourceMap": true,
+        "module": "es2015",
```
Comment (MattBidewell) [inline authorization2/tsconfig.json:9]: Try setting this to 2022

File: deploy.sh (hunk 1 above)
Comment (tal-nagra) [reply deploy.sh]: done

File: authorization2/tsconfig.json (hunk 3 above)
Comment (tal-nagra) [reply authorization2/tsconfig.json:9]: This has been changed to esnext

File: authorization2/package.json (hunk 2 above)
Comment (tal-nagra) [reply authorization2/package.json]: This has been amended to use crypto.randomUUID()

File: infrastructure/lambda/pre-merge-api.yaml [hunk 4]
```diff
@@ -33,7 +33,7 @@ paths:
         type: "aws_proxy"
         httpMethod: "POST"
         uri:
-          Fn::Sub: arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${AuthorizationFunction.Arn}/invocations
+          Fn::Sub: arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${AuthorizationFunctionTwo.Arn}/invocations
```
Comment (MattBidewell) [inline infrastructure/lambda/pre-merge-api.yaml:36]: todo undo these

**Review by jkunle — APPROVED** [review]

**Review by MattBidewell — APPROVED** [review]

**Pull request merged**